programs/drift/src/state/perp_market.rs

```diff
@@ -1689,3 +1696,152 @@ impl AMM {
+        CacheInfo {
+            position: 0i64,
+            slot: 0u64,
+            max_confidence_interval_multiplier: 1u64,
+            last_oracle_price_twap: 0i64,
+            oracle_price: 0i64,
+            oracle_confidence: 0u64,
+            oracle_delay: 0i64,
+            oracle_slot: 0u64,
+            _padding: [0u8; 7],
+            oracle: Pubkey::default(),
+            last_fee_pool_token_amount: 0u128,
+            last_net_pnl_pool_token_amount: 0i128,
+            last_settle_amount: 0u64,
+            last_settle_ts: 0i64,
+            oracle_source: 0u8,
+            quote_owed_from_lp: 0i64,
+        }
+    }
+}
+
+impl CacheInfo {
+    pub fn get_oracle_source(&self) -> DriftResult<OracleSource> {
+        Ok(OracleSource::try_from(self.oracle_source)?)
+    }
+
+    pub fn oracle_id(&self) -> DriftResult<OracleIdentifier> {
+        let oracle_source = self.get_oracle_source()?;
+        Ok((self.oracle, oracle_source))
+    }
+
+    pub fn get_last_available_amm_balance(&self) -> DriftResult<i128> {
+        let last_available_balance = self
+            .last_fee_pool_token_amount
+            .cast::<i128>()?
+            .safe_add(self.last_net_pnl_pool_token_amount)?;
+        Ok(last_available_balance)
+    }
+}
+
+#[zero_copy]
+#[derive(Default, Debug)]
+#[repr(C)]
+pub struct AmmCacheFixed {
+    pub bump: u8,
+    _pad: [u8; 3],
+    pub len: u32,
+}
+
+impl HasLen for AmmCacheFixed {
+    fn len(&self) -> u32 {
+        self.len
+    }
+}
+
+impl AmmCache {
+    pub fn space(num_markets: usize) -> usize {
+        8 + 8 + 4 + num_markets * CacheInfo::SIZE
+    }
+
+    pub fn validate(&self, state: &State) -> DriftResult<()> {
+        validate!(
+            self.cache.len() == state.number_of_markets as usize,
+            ErrorCode::DefaultError,
+            "Number of amm positions is different than number of markets"
+        )?;
+        Ok(())
+    }
+}
+
+impl_zero_copy_loader!(AmmCache, crate::id, AmmCacheFixed, CacheInfo);
+
+impl<'a> AccountZeroCopy<'a, CacheInfo, AmmCacheFixed> {
+    pub fn check_settle_staleness(&self, now: i64, threshold_ms: i64) -> DriftResult<()> {
+        for (i, cache_info) in self.iter().enumerate() {
+            if cache_info.last_settle_ts < now.saturating_sub(threshold_ms) {
+                msg!("AMM settle data is stale for perp market {}", i);
+                return Err(ErrorCode::AMMCacheStale.into());
+            }
+        }
+        Ok(())
+    }
+
+    pub fn check_perp_market_staleness(&self, slot: u64, threshold: u64) -> DriftResult<()> {
+        for (i, cache_info) in self.iter().enumerate() {
+            if cache_info.slot < slot.saturating_sub(threshold) {
+                msg!("Perp market cache info is stale for perp market {}", i);
+                return Err(ErrorCode::AMMCacheStale.into());
+            }
+        }
+        Ok(())
+    }
+
+    pub fn check_oracle_staleness(&self, slot: u64, threshold: u64) -> DriftResult<()> {
```

Comment:
did we consider having these computed and saved on the cache account so we dont have to keep calling it every time its needed?